File: scripts/find-top-teams.py
# script is used to filter out the teams that have been top 30 (for generating puzzles)
# prints out all teams that have been at least top 30 in some point in time
# deprecated since hltv blocks this now (use the js script instead)

from bs4 import BeautifulSoup
import csv
import time
import requests
import os
import ast
import json

def get_parsed_page(url):
  headers = {
    "referer": "https://www.hltv.org/stats",
    "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"
  }

  res = requests.get(url, headers=headers)

  return BeautifulSoup(res.text, 'html.parser')

THRESHOLDS = [30, 20, 10]

prev_all_teams_list = []
all_teams_path = os.path.join(os.path.dirname(__file__), '..', 'data', 'all-teams.txt')
with open(all_teams_path, 'r', encoding="utf8") as file:
  file_string = file.read()
  prev_all_teams_list = ast.literal_eval(file_string)
prev_all_teams = set(prev_all_teams_list)

prev_top_teams = {}
top_teams = {}

for THRESHOLD in THRESHOLDS:
  prev_top_teams_list = []

  top_teams_path = os.path.join(os.path.dirname(__file__), '..', 'data', 'top-' + str(THRESHOLD) + '-teams.txt')
  if os.path.exists(top_teams_path):
    # previous data exists, load it
    with open(top_teams_path, 'r', encoding="utf8") as file:
      file_string = file.read()
      prev_top_teams_list = ast.literal_eval(file_string)

  prev_top_teams[THRESHOLD] = set(prev_top_teams_list)
  top_teams[THRESHOLD] = set(prev_top_teams_list)

all_teams = set()

def read_data():
  with open(os.path.join(os.path.dirname(__file__), '..', 'data', 'playerData.csv'), 'r', encoding="utf8") as file:
    csvreader = csv.reader(file)
    top_row = next(csvreader)

    for row in csvreader:
      team_list = set(eval(row[18]))
      for team in team_list:
        all_teams.add(team)

read_data()

team_count = 0
for team in all_teams:
  print('loading team ' + team + ' (team ' + str(team_count+1) + ' of ' + str(len(all_teams)) + ')')
  team_count += 1

  if team in prev_top_teams[THRESHOLDS[-1]]:
    # team has already been processed, skip
    print('already processed team, skip')
    continue

  team_id = team.split('/')[0]

  team_page = get_parsed_page('https://www.hltv.org/team/' + team_id + '/a')

  if team_page.find('div', {'class': 'team-chart-container'}) is None:
    # whoops broken link, shouldnt happen ??
    print('skipping team')
    continue

  if len(team_page.find('div', {'class': 'team-chart-container'}).find_all('span', {'class': 'value'})) > 1:
    rank = int(team_page.find('div', {'class': 'team-chart-container'}).find_all('span', {'class': 'value'})[1].text[1:])

    for THRESHOLD in THRESHOLDS:
      if rank <= THRESHOLD:
        top_teams[THRESHOLD].add(team)

print('writing to all teams file')
f = open(all_teams_path, 'w', encoding="utf8")
f.write(json.dumps(list(all_teams)))
f.close()

for THRESHOLD in THRESHOLDS:
  print('writing to top ' + str(THRESHOLD) + ' teams file')
  write_to_top_team_path = os.path.join(os.path.dirname(__file__), '..', 'data', 'top-' + str(THRESHOLD) + '-teams.txt')
  f = open(write_to_top_team_path, 'w', encoding="utf8")
  f.write(json.dumps(list(top_teams[THRESHOLD])))
  f.close()

File: scripts/find-top-teams.js
// script is used to filter out the teams that have been top 30 (for generating puzzles)
// prints out all teams that have been at least top 30 in some point in time
const fs = require('fs').promises
const fsold = require('fs')
const csv = require('csv-parser')

const { loadBrowser, getParsedPage } = require('./retrieve-data-fns.js')

async function readCSV(allTeams) {
  return new Promise(async function (resolve, reject) {
    fsold.createReadStream('data/playerData.csv')
    .pipe(csv())
    .on('headers', (headers) => {

    })
    .on('data', (row) => {
      const rowData = Object.values(row)
      const teamList = JSON.parse(rowData[19])

      for (let i = 0; i < teamList.length; i++) {
        allTeams.add(teamList[i])
      }
    })
    .on('end', () => {
      resolve(0)
    })
  })
}

async function main() {
  const browserInfo = await loadBrowser();

  const THRESHOLDS = [30, 20, 10]

  const data = await fs.readFile('data/all-teams.txt', 'utf8')
  const prevAllTeamsList = JSON.parse(data)
  let prevAllTeams = new Set(prevAllTeamsList)

  const prevTopTeams = {}
  const topTeams = {}

  for (const THRESHOLD of THRESHOLDS) {
    const prevTopTeamsList = JSON.parse(await fs.readFile('data/top-' + THRESHOLD + '-teams.txt', 'utf8'))
    prevTopTeams[THRESHOLD] = new Set(prevTopTeamsList)
    topTeams[THRESHOLD] = new Set(prevTopTeamsList)
  }

  const allTeams = new Set()

  await readCSV(allTeams)

  let teamCount = 1
  for (const team of allTeams) {
    console.log('loading team ' + team + ' (team ' + teamCount + ' of ' + allTeams.size + ')')
    teamCount += 1

    if (prevTopTeams[THRESHOLDS[THRESHOLDS.length - 1]].has(team)) {
      // team has already been processed, skip
      console.log('already processed team, skip')
      continue
    }

    const teamId = team.split('/')[0]
    const teamPage = await getParsedPage(browserInfo, 'https://www.hltv.org/team/' + teamId + '/a', ['div', 'team-chart-container'])

    if (teamPage.find('div', {'class': 'team-chart-container'}) === undefined) {
      // whoops broken link, shouldnt happen ??
      console.log('skipping team')
      continue
    }

    if (teamPage.find('div', {'class': 'team-chart-container'}).findAll('span', {'class': 'value'}).length > 1) {
      const rank = parseInt(teamPage.find('div', {'class': 'team-chart-container'}).findAll('span', {'class': 'value'})[1].text.substring(1))

      for (const THRESHOLD of THRESHOLDS) {
        if (rank <= THRESHOLD) {
          topTeams[THRESHOLD].add(team)
        }
      }
    }
  }

  console.log('writing to all teams file')
  const allTeamsList = Array.from(allTeams)
  allTeamsList.sort()
  await fs.writeFile('data/all-teams.txt', JSON.stringify(allTeamsList))

  for (const THRESHOLD of THRESHOLDS) {
    console.log('writing to top ' + THRESHOLD + ' teams file')
    const topTeamsList = Array.from(topTeams[THRESHOLD])
    topTeamsList.sort()
    await fs.writeFile('data/top-' + THRESHOLD + '-teams.txt', JSON.stringify(topTeamsList))
  }

  await browserInfo.browser.close()
}

main()
